src/Google/Ads/GoogleAds/Lib/OAuth2TokenBuilder.php

```diff
@@ -155,23 +170,65 @@ public function withImpersonatedEmail(?string $impersonatedEmail): self
      *
      * @return FetchAuthTokenInterface the created OAuth2 object that can fetch auth tokens
      */
-    public function build()
+    public function build(): FetchAuthTokenInterface
     {
-        $this->defaultOptionals();
-        $this->validate();
+        // Determine the final scope array to use for User Refresh and ADC.
+        // // NOTE: $this->scopes is guaranteed to be set by defaultOptionals().
+        $scopeArrayForUserAndAdc = explode(' ', $this->scopes);
 
-        if ($this->jsonKeyFilePath !== null) {
+        // 1. Check for **EXPLICIT** Service Account Flow
+        if (!empty($this->jsonKeyFilePath)) {
+            if (is_null($this->scopes)) {
+                throw new InvalidArgumentException(
+                    "Both 'jsonKeyFilePath' and 'scopes' must be set when using service account flow."
+                );
+            }
+            if (!is_null($this->clientId) || !is_null($this->clientSecret) || !is_null($this->refreshToken)) {
+                throw new InvalidArgumentException(
+                    "Cannot have both service account flow and installed/web "
+                    . "application flow credential values set."
+                );
+            }
+            // Service Account flow uses the specific configured scope string
+            $scopesForExplicitFlows = explode(' ', $this->scopes);
```

Comment:
This and `$scopeArrayForUserAndAdc` look the same. You probably don't need so many variables.